File: packed5/packed5.go
// Package packed5 implements the Packed-5 string codec: a compact,
// self-delimiting representation for short strings dominated by ASCII letters,
// spaces, common punctuation, digits and Spanish accented characters.
//
//	import "github.com/ivanjoz/colbin/packed5"
//
//	buf := packed5.Append(nil, "helloWorld")
//
// # Units
//
// The alphabet is 32 five-bit codes, and **every token is a whole number of
// them**. That is the property the whole codec is built around: eight units are
// forty bits are five bytes exactly, so the packer builds one uint64 with fixed
// shifts, stores eight bytes and advances five. No accumulator, no variable
// drain, no token that straddles the grid at an offset the compiler cannot see.
//
// # Wire format
//
// A value is one frame. The first byte carries the header flags and, for all
// but the longest payloads, the payload length as well:
//
//	bit  0     PACKED_5             0 = raw payload, 1 = packed unit stream
//	bit  1     UPPERCASE            the stream starts in uppercase mode
//	bit  2     reserved             must be zero
//	bits 3-7   length code          payload byte length, or 31 = "see uvarint"
//
// A length code of 31 is followed by the payload length as an LEB128 uvarint.
// Codes 0..30 hold the length inline, so every frame whose payload fits in 30
// bytes costs exactly one byte of framing. The length is a *byte* count in both
// modes, which is what makes the frame self-delimiting.
//
// In raw mode the payload is the original byte string verbatim, valid UTF-8 or
// not. In packed mode the payload is the unit stream, five bits per unit,
// LSB-first within each byte, starting at bit 0 of the first payload byte.
//
// There is no pad prefix and no pad count. The unit count follows from the
// payload length alone:
//
//	units = len(payload) * 8 / 5
//
// because the encoder pads the stream to the unit grid with a trailing
// CASE_TOGGLE_SIMPLE. A simple toggle applies to the next letter, so one with no
// letter after it decodes to nothing: it is a no-op the alphabet already had.
// Padding with it costs no bytes — the rounding of 5*units bits up to a whole
// number of bytes always leaves room — and it is what lets the stream start on
// the grid at bit 0 rather than three bits into it.
//
// # Opcodes
//
//	0..25   letter a..z, cased by the current case mode
//	26      space
//	27      CASE_TOGGLE_SIMPLE   invert the case of the next letter only
//	28      CASE_TOGGLE_LONG     invert the case mode until the next 28
//	29      + 1 unit: index into symTable  (digits and common punctuation)
//	30      + 1 unit: index into extTable  (accents and rare punctuation),
//	                  or 31 to start a raw-byte escape
//	31      + 2 units: integer 0..1023, low five bits first
//
// # Raw escape
//
// Opcode 30 with operand 31 is followed by one unit holding n-1 for n in 1..4,
// then two units per raw byte — the low five bits, then the high three:
//
//	3 + 2n units
//
// The escape carries bytes, not runes. That makes the codec byte-exact for
// arbitrary input, including strings that are not valid UTF-8, and lets the
// encoder amortise the three-unit header over a run of up to four bytes instead
// of paying it per rune.
//
// # Encoder
//
// One greedy left-to-right pass, fused with the packer. There is no planning
// pass: the codec has no behavioural header flags left to price, because the
// case mode is carried by an ordinary CASE_TOGGLE_LONG that the encoder hoists
// into the header bit when it lands first. See encode.go.
//
// The encoder never enlarges a string. It writes into a buffer bounded by the
// raw length and falls back to the raw form the moment the packed stream would
// pass it, so the result is never longer than len(s) plus its framing.
package packed5

import "errors"

var (
	// ErrTruncated is returned when a frame ends before its declared payload,
	// or a token's operands run past the end of the unit stream.
	ErrTruncated = errors.New("colbin: packed5 string truncated")
	// ErrReservedSymbol is returned for extTable indices the current table
	// leaves unassigned.
	ErrReservedSymbol = errors.New("colbin: packed5 reserved symbol index")
	// ErrBadLength is returned when the frame's length prefix is malformed or
	// describes a payload larger than the buffer.
	ErrBadLength = errors.New("colbin: packed5 bad length prefix")
	// ErrBadEscape is returned for a raw escape whose count or byte halves fall
	// outside the ranges the encoder can produce.
	ErrBadEscape = errors.New("colbin: packed5 bad raw escape")
	// ErrBadHeader is returned when a reserved header bit is set.
	ErrBadHeader = errors.New("colbin: packed5 reserved header bit set")
)

// Header flag bits, in byte 0 of every frame.
const (
	flagPacked5   = 1 << 0
	flagUppercase = 1 << 1
	flagReserved  = 1 << 2

	lenShift = 3
	// lenInline is the largest payload length the header's five length bits can
	// hold; 31 is the escape code that defers to a uvarint.
	lenInline = 30
	lenEscape = 31
)

// The alphabet. Every code is one unit; the operands below are units too.
const (
	opSpace      = 26
	opCaseSimple = 27
	opCaseLong   = 28
	opSymbol     = 29
	opExt        = 30
	opNumber     = 31
)

// extEscape is the extTable operand that introduces a run of raw bytes.
const extEscape = 31

// maxEscapeRun is how many raw bytes one escape can carry; the count occupies
// one unit but only 1..4 are legal, so that the header stays worth amortising.
const maxEscapeRun = 4

// numberMax is the largest integer opcode 31 can carry in its two units.
const numberMax = 1023

// numberMaxDigits is the digit count of numberMax, and so the longest decimal
// run one number token can absorb.
const numberMaxDigits = 4

// symTable is the opcode 29 operand table: the characters a short record is
// actually made of once letters and spaces are accounted for. All 32 entries are
// assigned, so no operand of this opcode can be invalid.
//
// Digits sit here at two units rather than in a table of their own. A lone digit
// therefore costs ten bits where the pre-unit format charged nine — the one bit
// per digit that buying the grid costs, and less than the three bits per frame
// that dropping the pad prefix returns.
var symTable = [32]byte{
	'0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
	'.', ',', '-', '/', ':', ';', '_', '(', ')', '%',
	'#', '"', '\'', '!', '?', '@', '=', '+', '*', '&', '<', '>',
}

// extTable is the opcode 30 operand table: everything else worth a token.
// Index 31 is not a character, it is extEscape, and indices extReserved..30 are
// unassigned — rejected by the decoder today, so claiming them later is a clean
// format change rather than a silent reinterpretation.
//
// The accented characters are literals: the case mode does not apply to them,
// which is why both cases appear.
var extTable = [32]string{
	0: "ñ", 1: "á", 2: "é", 3: "í", 4: "ó", 5: "ú", 6: "ü",
	7: "Ñ", 8: "Á", 9: "É", 10: "Í", 11: "Ó", 12: "Ú",
	13: "€", 14: "$", 15: "~", 16: "`", 17: `\`, 18: "[", 19: "]",
	20: "^", 21: "{", 22: "}", 23: "|", 24: "\n", 25: "\t", 26: "\r", 27: "¿",
	28: "", 29: "", 30: "", // reserved
	31: "", // extEscape
}

// extReserved is the first unassigned extTable index.
const extReserved = 28

// asciiSym and asciiExt invert the two tables for their single-byte entries. A
// -1 means "this byte has no operand in that table".
//
// latinC2 and latinC3 do the same for the multi-byte entries, keyed on the low
// six bits of the UTF-8 continuation byte. Every non-ASCII entry above is Latin-1
// Supplement (a C2 or C3 lead) except '€', and continuation bytes run 0x80..0xBF,
// so within well-formed input those six bits identify the character outright —
// one array index instead of a rune decode and a table walk. Spanish text takes
// this path on nearly every word.
var (
	asciiSym [128]int8
	asciiExt [128]int8
	latinC2  [64]int8
	latinC3  [64]int8
	extEuro  int8 = -1
)

func init() {
	for i := range asciiSym {
		asciiSym[i] = -1
		asciiExt[i] = -1
	}
	for i := range latinC2 {
		latinC2[i] = -1
		latinC3[i] = -1
	}
	for i, c := range symTable {
		asciiSym[c] = int8(i)
	}
	for i, s := range extTable {
		if i >= extReserved {
			continue
		}
		switch {
		case len(s) == 1:
			asciiExt[s[0]] = int8(i)
		case len(s) == 2 && s[0] == 0xC2:
			latinC2[s[1]&0x3F] = int8(i)
		case len(s) == 2 && s[0] == 0xC3:
			latinC3[s[1]&0x3F] = int8(i)
		case s == "€":
			extEuro = int8(i)
		}
	}
}

// extMulti returns the extTable index of the multi-byte character at s[i] with
// its width in bytes, or -1 and 0.
func extMulti(s string, i int) (int8, int) {
	if i+1 >= len(s) || s[i+1]&0xC0 != 0x80 {
		return -1, 0
	}
	switch s[i] {
	case 0xC2:
		return latinC2[s[i+1]&0x3F], 2
	case 0xC3:
		return latinC3[s[i+1]&0x3F], 2
	case 0xE2:
		if i+2 < len(s) && s[i+1] == 0x82 && s[i+2] == 0xAC {
			return extEuro, 3
		}
	}
	return -1, 0
}

// payloadUnits is how many units a payload of n bytes holds. The encoder pads to
// the grid, so this is exact rather than an upper bound.
func payloadUnits(n int) int { return n * 8 / 5 }

// payloadBytes is how many bytes n units occupy.
func payloadBytes(n int) int { return (n*5 + 7) / 8 }

// uvarintLen is the LEB128 length of v, matching appendUvarint below.
func uvarintLen(v int) int {
	n := 1
	for v >= 0x80 {
		v >>= 7
		n++
	}
	return n
}

// appendUvarint writes v as LEB128. v is always a non-negative length here.
func appendUvarint(out []byte, v int) []byte {
	for v >= 0x80 {
		out = append(out, byte(v)|0x80)
		v >>= 7
	}
	return append(out, byte(v))
}

// readUvarint reads an LEB128 value from buf, returning it with the number of
// bytes consumed. It rejects overlong encodings and values that cannot be a
// length on this platform.
func readUvarint(buf []byte) (int, int, error) {
	var v uint64
	var shift uint
	for i, b := range buf {
		if i >= 9 {
			return 0, 0, ErrBadLength
		}
		v |= uint64(b&0x7F) << shift
		if b < 0x80 {
			if v > uint64(maxLen) {
				return 0, 0, ErrBadLength
			}
			return int(v), i + 1, nil
		}
		shift += 7
	}
	return 0, 0, ErrTruncated
}

// maxLen bounds a decoded payload length so that readUvarint cannot produce a
// value that overflows int on a 32-bit platform.
const maxLen = int(^uint32(0) >> 1)

// frameOverhead is the number of framing bytes a payload of n bytes needs: the
// header byte, plus a uvarint once the length outgrows the header's length bits.
func frameOverhead(n int) int {
	if n <= lenInline {
		return 1
	}
	return 1 + uvarintLen(n)
}

// isLetter reports whether c is an ASCII letter of either case.
//
// Setting bit 5 folds 'A'..'Z' onto 'a'..'z' and moves nothing else into that
// range, so one unsigned range test decides both cases at once.
func isLetter(c byte) bool { return (c|0x20)-'a' < 26 }

// isDigit reports whether c is an ASCII decimal digit. Unsigned wraparound turns
// the pair of comparisons into one.
func isDigit(c byte) bool { return c-'0' < 10 }

// isUpper reports the case of a byte already known to be an ASCII letter: for
// those, bit 5 is the case bit.
func isUpper(c byte) bool { return c&0x20 == 0 }

// letterIndex is the 0..25 alphabet position of an ASCII letter, either case.
func letterIndex(c byte) uint8 { return (c | 0x20) - 'a' }

File: rust/src/packed5.rs
//! The Packed-5 string codec, mirroring the Go `packed5` package: a compact,
//! self-delimiting representation for short strings dominated by ASCII letters,
//! spaces, common punctuation, digits and Spanish accented characters.
//!
//! ```
//! use colbin::packed5;
//!
//! let mut buf = Vec::new();
//! packed5::append(&mut buf, b"el nino comio jamon");
//! let (bytes, read) = packed5::decode(&buf)?;
//! assert_eq!(bytes, b"el nino comio jamon");
//! assert_eq!(read, buf.len());
//! # Ok::<(), colbin::Error>(())
//! ```
//!
//! It is an opt-in encoding rather than the default: a raw blob is a sub-slice
//! of the message on the way out and a memcpy on the way in, where a packed one
//! is a pass over every character on both sides. A `BLOB` descriptor carries the
//! choice per field, so a message can hold both and a reader is never told which
//! to expect — see [`crate::wire::Writer8::packed_string`].
//!
//! # Units
//!
//! The alphabet is 32 five-bit codes and **every token is a whole number of
//! them**. Eight units are forty bits are five bytes exactly, so the packer
//! builds one `u64` with fixed shifts, writes eight bytes and advances five,
//! and the reader is the mirror. Nothing straddles the byte grid.
//!
//! # Wire format
//!
//! Byte 0 carries the header flags and, for all but the longest payloads, the
//! payload length as well:
//!
//! ```text
//! bit  0     PACKED_5     0 = raw payload, 1 = packed unit stream
//! bit  1     UPPERCASE    the stream starts in uppercase mode
//! bit  2     reserved     must be zero
//! bits 3-7   length code  payload byte length, or 31 = "see uvarint"
//! ```
//!
//! In packed mode the payload is the unit stream, five bits per unit, LSB-first
//! within each byte, starting at bit 0 of the first payload byte. There is no
//! pad prefix: the unit count is `payload.len() * 8 / 5`, because the encoder
//! pads to the grid with a trailing `CASE_TOGGLE_SIMPLE`, which applies to no
//! letter and so decodes to nothing.
//!
//! ```text
//! 0..25   letter a..z, cased by the current case mode      1 unit
//! 26      space                                            1 unit
//! 27      CASE_TOGGLE_SIMPLE   the next letter only         1 unit
//! 28      CASE_TOGGLE_LONG     until the next one           1 unit
//! 29      + symbol table index                             2 units
//! 30      + extension table index, or 31 to escape         2 units
//! 31      + a 10-bit integer, low five bits first          3 units
//! ```
//!
//! A raw escape is `30, 31`, one unit holding `n-1` for `n` in 1..=4, then two
//! units per byte — the low five bits, then the high three. It carries **bytes**,
//! not runes, which is what makes the codec byte exact for input that is not
//! valid UTF-8 — and why this module works in `[u8]` and leaves the caller to
//! decide what a frame that is not UTF-8 means.

use crate::error::Packed5Error;

// Header flag bits, in byte 0 of every frame.
const FLAG_PACKED5: u8 = 1 << 0;
const FLAG_UPPERCASE: u8 = 1 << 1;
const FLAG_RESERVED: u8 = 1 << 2;

const LEN_SHIFT: u8 = 3;
const LEN_INLINE: usize = 30;
const LEN_ESCAPE: usize = 31;

// The alphabet.
const OP_SPACE: u8 = 26;
const OP_CASE_SIMPLE: u8 = 27;
const OP_CASE_LONG: u8 = 28;
const OP_SYMBOL: u8 = 29;
const OP_EXT: u8 = 30;
const OP_NUMBER: u8 = 31;

/// The extension table operand that introduces a run of raw bytes.
const EXT_ESCAPE: u8 = 31;
/// The first unassigned extension table index.
const EXT_RESERVED: u8 = 28;

const MAX_ESCAPE_RUN: usize = 4;
const NUMBER_MAX: u32 = 1023;
const NUMBER_MAX_DIGITS: usize = 4;

/// The opcode 29 operand table: what a short record is made of once letters and
/// spaces are accounted for. All 32 entries are assigned.
const SYM_TABLE: [u8; 32] = [
    b'0', b'1', b'2', b'3', b'4', b'5', b'6', b'7', b'8', b'9', b'.', b',', b'-', b'/', b':', b';',
    b'_', b'(', b')', b'%', b'#', b'"', b'\'', b'!', b'?', b'@', b'=', b'+', b'*', b'&', b'<', b'>',
];

/// The opcode 30 operand table. Indices 28..=30 are reserved and 31 is the
/// escape, so a decoder rejects them rather than reinterpreting them later.
const EXT_TABLE: [&str; 28] = [
    "ñ", "á", "é", "í", "ó", "ú", "ü", "Ñ", "Á", "É", "Í", "Ó", "Ú", "€", "$", "~", "`", "\\", "[",
    "]", "^", "{", "}", "|", "\n", "\t", "\r", "¿",
];

/// The multi-byte entries of `EXT_TABLE`, matched on their UTF-8 bytes.
const EXT_MULTI: [(&[u8], u8); 15] = [
    ("ñ".as_bytes(), 0),
    ("á".as_bytes(), 1),
    ("é".as_bytes(), 2),
    ("í".as_bytes(), 3),
    ("ó".as_bytes(), 4),
    ("ú".as_bytes(), 5),
    ("ü".as_bytes(), 6),
    ("Ñ".as_bytes(), 7),
    ("Á".as_bytes(), 8),
    ("É".as_bytes(), 9),
    ("Í".as_bytes(), 10),
    ("Ó".as_bytes(), 11),
    ("Ú".as_bytes(), 12),
    ("€".as_bytes(), 13),
    ("¿".as_bytes(), 27),
];

fn is_letter(c: u8) -> bool {
    (c | 0x20).wrapping_sub(b'a') < 26
}

fn is_digit(c: u8) -> bool {
    c.wrapping_sub(b'0') < 10
}

fn is_upper(c: u8) -> bool {
    c & 0x20 == 0
}

fn letter_index(c: u8) -> u8 {
    (c | 0x20) - b'a'
}

/// The symbol table index of an ASCII byte, or `None`.
fn ascii_sym(c: u8) -> Option<u8> {
    SYM_TABLE.iter().position(|&s| s == c).map(|i| i as u8)
}

/// The extension table index of a single-byte entry, or `None`.
fn ascii_ext(c: u8) -> Option<u8> {
    if c >= 0x80 {
        return None;
    }
    EXT_TABLE
        .iter()
        .position(|s| s.as_bytes() == [c])
        .map(|i| i as u8)
}

/// The extension table index of the multi-byte character at `bytes[at]`, with
/// its width.
fn ext_multi(bytes: &[u8], at: usize) -> Option<(u8, usize)> {
    for (utf8, index) in EXT_MULTI {
        if bytes.len() - at >= utf8.len() && &bytes[at..at + utf8.len()] == utf8 {
            return Some((index, utf8.len()));
        }
    }
    None
}

/// Whether the byte at `at` has no token of its own.
fn escapes(bytes: &[u8], at: usize) -> bool {
    let c = bytes[at];
    if is_letter(c) || c == b' ' {
        return false;
    }
    if c < 0x80 {
        return ascii_sym(c).is_none() && ascii_ext(c).is_none();
    }
    ext_multi(bytes, at).is_none()
}

/// How many units a payload of `n` bytes holds. The encoder pads to the grid, so
/// this is exact rather than an upper bound.
fn payload_units(n: usize) -> usize {
    n * 8 / 5
}

/// How many bytes `n` units occupy.
fn payload_bytes(n: usize) -> usize {
    (n * 5).div_ceil(8)
}

/// Whether the stream should open in uppercase: the first letter's case.
///
/// It is never worse than starting lower. A leading run of `k` opposite-case
/// letters costs `2k` units from lower for `k` of one or two and `1+k` from
/// three up, where starting in that mode costs `k` plus the one toggle that
/// returns — equal at `k>=3` and strictly better below it.
fn opens_upper(bytes: &[u8]) -> bool {
    bytes
        .iter()
        .find(|&&c| is_letter(c))
        .is_some_and(|&c| is_upper(c))
}

/// Packs units into a byte buffer, eight at a time.
struct Writer {
    out: Vec<u8>,
    acc: u64,
    pending: u8,
    units: usize,
}

impl Writer {
    fn new(capacity: usize) -> Self {
        Self {
            out: Vec::with_capacity(capacity),
            acc: 0,
            pending: 0,
            units: 0,
        }
    }

    fn put(&mut self, unit: u8) {
        self.acc |= u64::from(unit) << (5 * self.pending);
        self.pending += 1;
        self.units += 1;
        if self.pending == 8 {
            self.out.extend_from_slice(&self.acc.to_le_bytes()[..5]);
            self.acc = 0;
            self.pending = 0;
        }
    }

    /// Flushes the partial group and pads the stream to the unit grid.
    fn finish(mut self) -> Vec<u8> {
        if payload_units(payload_bytes(self.units)) > self.units {
            self.put(OP_CASE_SIMPLE);
        }
        if self.pending > 0 {
            let width = payload_bytes(usize::from(self.pending));
            self.out.extend_from_slice(&self.acc.to_le_bytes()[..width]);
        }
        self.out
    }
}

/// The greedy left-to-right scan, fused with the packer. There is no planning
/// pass: the format has no behavioural header flags left to price.
fn tokenize(w: &mut Writer, bytes: &[u8], upper: bool) {
    let mut cur = upper;
    let mut at = 0;
    while at < bytes.len() {
        let c = bytes[at];
        if is_letter(c) {
            let index = letter_index(c);
            if is_upper(c) == cur {
                w.put(index);
                at += 1;
                continue;
            }
            // Two simple toggles cost what a pair of long ones does, so the long
            // form only pays from a run of three.
            let mut run = 0;
            let mut scan = at;
            while scan < bytes.len() && is_letter(bytes[scan]) && is_upper(bytes[scan]) != cur {
                run += 1;
                scan += 1;
            }
            if run >= 3 {
                w.put(OP_CASE_LONG);
                cur = !cur;
                continue; // re-read the letter, now in the matching mode
            }
            w.put(OP_CASE_SIMPLE);
            w.put(index);
            at += 1;
        } else if c == b' ' {
            w.put(OP_SPACE);
            at += 1;
        } else if is_digit(c) {
            // The longest prefix ten bits can hold. A token may not carry a
            // leading zero, since it decodes as a plain decimal integer and
            // "00123" must not come back as "123"; a lone "0" is fine.
            let (mut value, mut best, mut best_len) = (0u32, 0u32, 0usize);
            for len in 1..=NUMBER_MAX_DIGITS {
                if at + len > bytes.len() {
                    break;
                }
                let digit = bytes[at + len - 1];
                if !is_digit(digit) || (len > 1 && bytes[at] == b'0') {
                    break;
                }
                value = value * 10 + u32::from(digit - b'0');
                if value > NUMBER_MAX {
                    break;
                }
                best = value;
                best_len = len;
            }
            if best_len == 1 {
                w.put(OP_SYMBOL);
                w.put(c - b'0');
            } else {
                w.put(OP_NUMBER);
                w.put((best & 31) as u8);
                w.put((best >> 5) as u8);
            }
            at += best_len;
        } else if let Some(index) = ascii_sym(c) {
            w.put(OP_SYMBOL);
            w.put(index);
            at += 1;
        } else if let Some(index) = ascii_ext(c) {
            w.put(OP_EXT);
            w.put(index);
            at += 1;
        } else if let Some((index, width)) = ext_multi(bytes, at) {
            w.put(OP_EXT);
            w.put(index);
            at += width;
        } else {
            let mut run = 1;
            while run < MAX_ESCAPE_RUN && at + run < bytes.len() && escapes(bytes, at + run) {
                run += 1;
            }
            w.put(OP_EXT);
            w.put(EXT_ESCAPE);
            w.put((run - 1) as u8);
            for k in 0..run {
                let b = bytes[at + k];
                w.put(b & 31);
                w.put(b >> 5);
            }
            at += run;
        }
    }
}

/// The packed unit stream for `bytes`, with the case mode it opens in, or `None`
/// when the packed form would not be smaller than the raw bytes.
///
/// This is the form an embedded string takes: a container that already carries a
/// length and a place to record the case mode — a `BLOB` descriptor does both —
/// needs no frame header of its own.
pub fn payload(bytes: &[u8]) -> Option<(Vec<u8>, bool)> {
    if bytes.is_empty() {
        return None;
    }
    let upper = opens_upper(bytes);
    let mut w = Writer::new(bytes.len() + 8);
    tokenize(&mut w, bytes, upper);
    let out = w.finish();
    if out.len() >= bytes.len() {
        return None;
    }
    Some((out, upper))
}

/// Appends one self-delimiting frame for `bytes`.
///
/// The packed form is used only when its frame is strictly smaller, so this
/// never enlarges a string.
pub fn append(out: &mut Vec<u8>, bytes: &[u8]) {
    if bytes.is_empty() {
        out.push(0); // raw, empty payload
        return;
    }
    match payload(bytes) {
        Some((stream, upper)) => {
            let mut flags = FLAG_PACKED5;
            if upper {
                flags |= FLAG_UPPERCASE;
            }
            append_header(out, flags, stream.len());
            out.extend_from_slice(&stream);
        }
        None => {
            append_header(out, 0, bytes.len());
            out.extend_from_slice(bytes);
        }
    }
}

/// The number of bytes [`append`] would add for `bytes`.
pub fn size(bytes: &[u8]) -> usize {
    if bytes.is_empty() {
        return 1;
    }
    let raw = frame_overhead(bytes.len()) + bytes.len();
    match payload(bytes) {
        Some((stream, _)) => {
            let packed = frame_overhead(stream.len()) + stream.len();
            if packed < raw { packed } else { raw }
        }
        None => raw,
    }
}

fn append_header(out: &mut Vec<u8>, flags: u8, payload_len: usize) {
    if payload_len <= LEN_INLINE {
        out.push(flags | (payload_len as u8) << LEN_SHIFT);
        return;
    }
    out.push(flags | (LEN_ESCAPE as u8) << LEN_SHIFT);
    append_uvarint(out, payload_len);
}

fn append_uvarint(out: &mut Vec<u8>, mut value: usize) {
    while value >= 0x80 {
        out.push((value as u8) | 0x80);
        value >>= 7;
    }
    out.push(value as u8);
}

fn uvarint_len(mut value: usize) -> usize {
    let mut n = 1;
    while value >= 0x80 {
        value >>= 7;
        n += 1;
    }
    n
}

fn frame_overhead(n: usize) -> usize {
    if n <= LEN_INLINE {
        1
    } else {
        1 + uvarint_len(n)
    }
}

fn read_uvarint(buf: &[u8]) -> Result<(usize, usize), Packed5Error> {
    let mut value: u64 = 0;
    let mut shift = 0;
    for (i, &b) in buf.iter().enumerate() {
        if i >= 9 {
            return Err(Packed5Error::BadLength);
        }
        value |= u64::from(b & 0x7F) << shift;
        if b < 0x80 {
            if value > u64::from(u32::MAX >> 1) {
                return Err(Packed5Error::BadLength);
            }
            return Ok((value as usize, i + 1));
        }
        shift += 7;
    }
    Err(Packed5Error::Truncated)
}

/// One parsed frame prefix.
struct Frame<'a> {
    payload: &'a [u8],
    total: usize,
    packed: bool,
    upper: bool,
}

fn frame(buf: &[u8]) -> Result<Frame<'_>, Packed5Error> {
    let &header = buf.first().ok_or(Packed5Error::Truncated)?;
    if header & FLAG_RESERVED != 0 {
        return Err(Packed5Error::BadHeader);
    }
    let mut at = 1;
    let mut length = usize::from(header >> LEN_SHIFT);
    if length == LEN_ESCAPE {
        let (value, read) = read_uvarint(&buf[at..])?;
        // The escape must not re-encode a length the header could have held; one
        // length has one encoding, so a frame has one byte representation.
        if value <= LEN_INLINE {
            return Err(Packed5Error::BadLength);
        }
        length = value;
        at += read;
    }
    if length > buf.len() - at {
        return Err(Packed5Error::Truncated);
    }
    Ok(Frame {
        payload: &buf[at..at + length],
        total: at + length,
        packed: header & FLAG_PACKED5 != 0,
        upper: header & FLAG_UPPERCASE != 0,
    })
}

/// Unpacks the payload into units, eight at a time.
fn units_of(payload: &[u8]) -> Vec<u8> {
    let count = payload_units(payload.len());
    let mut units = Vec::with_capacity(count);
    let mut at = 0;
    while units.len() < count {
        let mut word = [0u8; 8];
        let take = (payload.len() - at).min(8);
        word[..take].copy_from_slice(&payload[at..at + take]);
        let value = u64::from_le_bytes(word);
        for k in 0..8 {
            if units.len() == count {
                break;
            }
            units.push(((value >> (5 * k)) & 31) as u8);
        }
        at += 5;
    }
    units
}

/// Walks a bare unit stream — a payload with no frame header, as [`payload`]
/// writes it — and appends the bytes it holds to `dst`.
pub fn append_string(
    dst: &mut Vec<u8>,
    payload: &[u8],
    upper: bool,
) -> Result<(), Packed5Error> {
    let units = units_of(payload);
    let mut cur = upper;
    let mut pending = false;
    let mut at = 0;
    while at < units.len() {
        let op = units[at];
        at += 1;
        match op {
            0..=25 => {
                // Uppercase exactly when the two disagree.
                dst.push(if cur != pending { b'A' } else { b'a' } + op);
                pending = false;
            }
            OP_SPACE => dst.push(b' '),
            OP_CASE_SIMPLE => pending = true,
            OP_CASE_LONG => cur = !cur,
            OP_SYMBOL => {
                let &index = units.get(at).ok_or(Packed5Error::Truncated)?;
                at += 1;
                dst.push(SYM_TABLE[usize::from(index)]);
            }
            OP_EXT => {
                let &index = units.get(at).ok_or(Packed5Error::Truncated)?;
                at += 1;
                if index != EXT_ESCAPE {
                    if index >= EXT_RESERVED {
                        return Err(Packed5Error::ReservedSymbol);
                    }
                    dst.extend_from_slice(EXT_TABLE[usize::from(index)].as_bytes());
                    continue;
                }
                let &count = units.get(at).ok_or(Packed5Error::Truncated)?;
                at += 1;
                let count = usize::from(count) + 1;
                if count > MAX_ESCAPE_RUN {
                    return Err(Packed5Error::BadEscape);
                }
                if at + 2 * count > units.len() {
                    return Err(Packed5Error::Truncated);
                }
                for _ in 0..count {
                    let (low, high) = (units[at], units[at + 1]);
                    if high > 7 {
                        return Err(Packed5Error::BadEscape); // a byte is five bits and three
                    }
                    dst.push(low | high << 5);
                    at += 2;
                }
            }
            _ => {
                if at + 2 > units.len() {
                    return Err(Packed5Error::Truncated);
                }
                let value = u32::from(units[at]) | u32::from(units[at + 1]) << 5;
                at += 2;
                let mut buf = [0u8; 4];
                let mut len = 0;
                let mut v = value;
                loop {
                    buf[len] = b'0' + (v % 10) as u8;
                    len += 1;
                    v /= 10;
                    if v == 0 {
                        break;
                    }
                }
                for k in (0..len).rev() {
                    dst.push(buf[k]);
                }
            }
        }
    }
    Ok(())
}

/// Reads one frame from the front of `buf`, returning its bytes and the number
/// of bytes consumed, so frames can be read back to back.
pub fn decode(buf: &[u8]) -> Result<(Vec<u8>, usize), Packed5Error> {
    let mut out = Vec::new();
    let read = append_decoded(&mut out, buf)?;
    Ok((out, read))
}

/// [`decode`] appending onto a caller's buffer, for a run of frames gathered
/// into one backing array.
pub fn append_decoded(dst: &mut Vec<u8>, buf: &[u8]) -> Result<usize, Packed5Error> {
    let frame = frame(buf)?;
    if !frame.packed {
        dst.extend_from_slice(frame.payload);
        return Ok(frame.total);
    }
    append_string(dst, frame.payload, frame.upper)?;
    Ok(frame.total)
}
